src/Microsoft.ML.FastTree/GamModelParameters.cs

```diff
@@ -0,0 +1,911 @@
+                        // Add the value and subtract the value at zero that was previously accounted for
+                        value += GetBinEffect(j, featuresValues[i]) - GetBinEffect(j, 0);
+                }
+            }
+
+            response = (float)value;
+        }
+
+        internal double GetFeatureBinsAndScore(in VBuffer<float> features, int[] bins)
+        {
+            Host.CheckParam(features.Length == _inputLength, nameof(features));
+            Host.CheckParam(Utils.Size(bins) == _numFeatures, nameof(bins));
+
+            double value = Intercept;
+            var featuresValues = features.GetValues();
+            if (features.IsDense)
+            {
+                for (int i = 0; i < featuresValues.Length; ++i)
+                {
+                    if (_inputFeatureToDatasetFeatureMap.TryGetValue(i, out int j))
+                        value += GetBinEffect(j, featuresValues[i], out bins[j]);
+                }
+            }
+            else
+            {
+                var featuresIndices = features.GetIndices();
+                // Add in the precomputed results for all features
+                value += _valueAtAllZero;
+                Array.Copy(_binsAtAllZero, bins, _numFeatures);
+
+                // Update the results for features we have
+                for (int i = 0; i < featuresValues.Length; ++i)
+                {
+                    if (_inputFeatureToDatasetFeatureMap.TryGetValue(featuresIndices[i], out int j))
+                        // Add the value and subtract the value at zero that was previously accounted for
+                        value += GetBinEffect(j, featuresValues[i], out bins[j]) - GetBinEffect(j, 0);
+                }
+            }
+            return value;
+        }
+
+        private double GetBinEffect(int featureIndex, double featureValue)
+        {
+            Contracts.Assert(0 <= featureIndex && featureIndex < _numFeatures);
+            int index = Algorithms.FindFirstGE(_binUpperBounds[featureIndex], featureValue);
+            return _binEffects[featureIndex][index];
+        }
+
+        private double GetBinEffect(int featureIndex, double featureValue, out int binIndex)
+        {
+            Contracts.Assert(0 <= featureIndex && featureIndex < _numFeatures);
+            binIndex = Algorithms.FindFirstGE(_binUpperBounds[featureIndex], featureValue);
+            return _binEffects[featureIndex][binIndex];
+        }
+
+        /// <summary>
+        /// Get the bin upper bounds for each feature.
+        /// </summary>
+        /// <param name="featureIndex">The index of the feature (in the training vector) to get.</param>
+        /// <returns>The bin upper bounds. May be null if this feature has no bins.</returns>
+        public double[] GetFeatureBinUpperBounds(int featureIndex)
+        {
+            Contracts.Assert(0 <= featureIndex && featureIndex < _numFeatures);
```

Comment:
>Assert [](start = 22, length = 6)

use Check for public methods.
Also give it proper message.

Reply:
Done.